lib/spack/spack/directives.py

```diff
@@ -341,16 +347,28 @@ def _execute_version(pkg):
 
             kwargs["checksum"] = checksum
 
-        # Store kwargs for the package to later with a fetch_strategy.
+        if isinstance(ver, float):
+            raise VersionError(
+                f"{pkg.name}: declared version '{ver}' in package should be a string. "
+                "floating point numbers are ambiguous."
+            )
+        elif not isinstance(ver, (int, str)):
+            raise VersionError(
+                f"{pkg.name}: declared version '{ver!r}' in package should be a string or int."
+            )
```

Comment:
I feel like we should push people towards strings here, as ints can become floats easily with casual editing.  Right now one message says strings and the other says strings OR ints.

I see there are some packages that call `Version(int)` but I'd rather fix that in `develop`, deprecate ints entirely, and remove the int option in 0.21.

Reply:
That's done